### PR TITLE
MEF2 throws a System.TypeLoadException

### PR DESCRIPTION
**Pull request by DevInstinct**
Make sure the typeArguments follow the generic type parameters
constraints.
If not, then the part is not a match and must be ignored.
Otherwise, _partType.MakeGenericType throws a System.TypeLoadException.
(Unlike MEF2, MEF on the full .NET framework filters correctly and
ignores parts that do not match the constraints, so this fix makes the
behavior the same in both versions).

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @dsplaisted

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs [hunk 1]
```diff
@@ -191,6 +191,21 @@ public CompositeActivator GetActivator(DependencyAccessor definitionAccessor, IE
 
         public bool TryCloseGenericPart(Type[] typeArguments, out DiscoveredPart closed)
         {
+            // DevInstinct Fix Begin
```
Comment (terrajobst) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: Could you remove these markers?

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs [hunk 2]
```diff
@@ -191,6 +191,21 @@ public CompositeActivator GetActivator(DependencyAccessor definitionAccessor, IE
...
+            // Make sure the typeArguments follow the generic type parameters constraints.
+            // If not, then the part is not a match and must be ignored.
+            // Otherwise, _partType.MakeGenericType throws a System.TypeLoadException.
+            // (Unlike MEF2, MEF on the full .NET framework filters correctly and ignores parts that do not match the constraints, 
+            // so this fix makes the behavior the same in both versions).
+            for (int index = 0; index < _partType.GenericTypeParameters.Length; index++)
+                foreach (var genericParameterConstraints in _partType.GenericTypeParameters[index].GetTypeInfo().GetGenericParameterConstraints())
+                    if (!genericParameterConstraints.GetTypeInfo().IsAssignableFrom(typeArguments[index].GetTypeInfo()))
```
Comment (terrajobst) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: To make it more readable, we should add braces to the two loops.

**Comment by terrajobst** [comment]
Hey @DevInstinct, thanks for submitting the PR! I've added a few comments regarding style. But I'll defer to @dsplaisted for the code review regarding correctness.

**Comment by dsplaisted** [comment]
I took a glance at this change and it seems to make sense.  However, I'd like to understand more about when this would happen.  Under what scenario would a list of types be passed in that don't match the generic type constraints?  Can we add test coverage for this?

**Comment by Petermarcu** [comment]
@davkean

**Comment by DevInstinct** [comment]
Hi,

The exception happens when you have an interface inheritance chain. The exception type may vary depending on the way the code and composition is done. In this simplified example, MEF will throw ArgumentException [GenericArguments[0], ICar', on 'BookHandler`1[T]' violates the constraint of type 'T'] when trying to compose the CarHandlers because without the proposed fix MEF tries to include the BookHandler in the CarHandlers enumeration.

The fix adds a check on the constraint and, in this example, would detect that the constraint (where T : IBook) is not satisfied for the BookHandler and exclude that handler from the composition for the CarHandlers.

The standard version of MEF already supports the scenario properly - I got the exception when I moved my code to the lightweight version of MEF.

```
public interface IThing { }
public interface ICar : IThing { }
public interface IBook : IThing { }
public interface IPaperBook : IBook { }
public interface IDigitalBook : IBook { }

public interface IHandler<T> where T : IThing { }

[Export(typeof(IHandler<>))]
public class ThingHandler<T> : IHandler<T>
    where T : IThing
{
}

[Export(typeof(IHandler<>))]
public class BookHandler<T> : IHandler<T>
    where T : IBook
{
}

[Export]
public class CarComposition
{
    [ImportMany]
    public IEnumerable<IHandler<ICar>> CarHandlers { get; set; }
}
```

If you still need test coverage for this in the project, can you please tell me where to add it. I'm not sure where it would fit in the solution. 
Thanks,
Martin

**Comment by karelz** [comment]
@dsplaisted can you please follow up?

**Comment by Petermarcu** [comment]
@nblumhardt, any input on this one?

**Comment by nblumhardt** [comment]
@Petermarcu looks good to me :+1: 

@DevInstinct the tests for this assembly are in: https://github.com/dotnet/corefx/tree/master/src/System.Composition/tests

**Comment by Petermarcu** [comment]
Thanks @nblumhardt , @DevInstinct , can you add some test coverage and we can get this merged?

**Comment by Petermarcu** [comment]
@DevInstinct. just checking in on whether you'll be able to add some tests so we can get this merged?

**Comment by DevInstinct** [comment]
Hi, I took a few hours today to check the test project @nblumhardt mentionned, and I can't get my environment straight (compilations and warnings about unresolved test-runtime for .NET Standard v1.3 and nuget package downgrade). It may be because some time have passed since I branched to do this pull request. I'm using the solution in src\System.Composition with Visual Studio 2015 Update 3.
Not quite familiar with Github either since it's my first pull request. Would you have some advice to help get me running and maybe update this pull request to a source version I can work with? Thanks.

**Comment by karelz** [comment]
@dsplaisted @stephentoub can you please advise here?

**Comment by Petermarcu** [comment]
@DevInstinct, have you pulled upstream into your master branch to get up with the latest changes? If you set your upstream to dotnet/corefx and pull local, then push to your master on github, everything should get back in sync.

**Comment by DevInstinct** [comment]
Ok, I did the sync and then added the tests. It should be good now.

**Comment by Petermarcu** [comment]
Cool, the last thing you'll need to do assuming all the tests pass and folks sign off on the test code is to squash the commits so we don't have the merge commits and the whole change only adds one line in the history. Let us know if you need help with that. Thanks for everything so far!

**Comment by Petermarcu** [comment]
Nevermind, @stephentoub showed me that we can do that for you on merge :)

**Comment by Petermarcu** [comment]
@nblumhardt , tests look good?

**Comment by nblumhardt** [comment]
LGTM 👍

Thinking back now, I believe the reason we didn't tackle this was the difficulties inherent in checking more complex constraints like:

```csharp
class AResponder<T,U> : IResponder<T,U> where T: U {}
```

I don't think failing in these cases is a reason not to proceed with this improvement, but good to keep in mind if/when future bug reports pop up. (If there is a straightforward way to lift this limitation it would be great to cover all cases, but I'm not aware of any current API that can check these rigorously.)

**Pull request merged**